File: backend/Modules/Envios/EnvioService.cs
using CodeMX.Api.Modules.Evaluacion;

namespace CodeMX.Api.Modules.Envios;

/// <summary>
/// Implementación del módulo Envíos. Coordina el flujo de un envío:
///   1. Guarda el envío como PENDIENTE.
///   2. Pide la evaluación al módulo Evaluación (IEvaluacionApi).
///   3. Actualiza el veredicto.
///   4. Si es el primer ACEPTADO del usuario en ese reto, <b>publica el evento</b> a los
///      observadores registrados (patrón Observer). Envíos ya no conoce a Ranking: solo
///      notifica; quien quiera reaccionar se suscribe como IEnvioObserver.
/// </summary>
public class EnvioService : IEnviosApi
{
    private readonly EnvioRepository _repo;
    private readonly IEvaluacionApi _evaluacion;
    private readonly IEnumerable<IEnvioObserver> _observers;

    public EnvioService(EnvioRepository repo, IEvaluacionApi evaluacion, IEnumerable<IEnvioObserver> observers)
    {
        _repo = repo;
        _evaluacion = evaluacion;
        _observers = observers;
    }

    public async Task<Envio> EnviarAsync(Envio envio)
    {
        envio.Veredicto = Veredicto.PENDIENTE;
        var guardado = await _repo.Guardar(envio);

        try
        {
            var solicitud = new SolicitudEvaluacion(envio.RetoId, envio.CodigoFuente);
            var resultado = await _evaluacion.EvaluarAsync(solicitud);

            var veredicto = Enum.Parse<Veredicto>(resultado.Veredicto);

            bool esPrimerAceptado = veredicto == Veredicto.ACEPTADO
                && !await _repo.TieneAceptado(envio.UsuarioId, envio.RetoId);

            guardado.Veredicto = veredicto;
            guardado = await _repo.Guardar(guardado);

            if (esPrimerAceptado)
            {
                var evento = new EnvioAceptadoEvent(envio.UsuarioId, envio.RetoId);
                foreach (var observer in _observers)
                    await observer.OnEnvioAceptadoAsync(evento);
            }
        }
        catch
        {
            guardado.Veredicto = Veredicto.ERROR_EN_EJECUCION;
            await _repo.Guardar(guardado);
        }

        return guardado;
    }

    public Task<Envio?> BuscarPorIdAsync(long id) => _repo.BuscarPorId(id);

    public Task<List<Envio>> ListarPorUsuarioAsync(long usuarioId) => _repo.ListarPorUsuario(usuarioId);

    public Task<List<Envio>> ListarPorRetoAsync(long retoId) => _repo.ListarPorReto(retoId);

    public Task<bool> TieneAceptadoAsync(long usuarioId, long retoId) => _repo.TieneAceptado(usuarioId, retoId);
}

File: backend-java/src/main/java/mx/codemx/modules/envios/EnvioService.java
package mx.codemx.modules.envios;

import java.util.List;
import java.util.Optional;
import mx.codemx.modules.evaluacion.EvaluacionApi;
import mx.codemx.modules.evaluacion.ResultadoEvaluacion;
import mx.codemx.modules.evaluacion.SolicitudEvaluacion;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
public class EnvioService implements EnviosApi {

    private final EnvioRepository repo;
    private final EvaluacionApi evaluacion;
    private final List<EnvioObserver> observers;

    public EnvioService(EnvioRepository repo, EvaluacionApi evaluacion, List<EnvioObserver> observers) {
        this.repo = repo;
        this.evaluacion = evaluacion;
        this.observers = observers;
    }

    @Override
    public Envio enviar(Envio envio) {
        envio.setVeredicto(Veredicto.PENDIENTE);
        Envio guardado = repo.save(envio);

        try {
            SolicitudEvaluacion solicitud =
                    new SolicitudEvaluacion(envio.getRetoId(), envio.getCodigoFuente());
            ResultadoEvaluacion resultado = evaluacion.evaluar(solicitud);

            Veredicto veredicto = Veredicto.valueOf(resultado.veredicto());

            boolean esPrimerAceptado = veredicto == Veredicto.ACEPTADO
                    && !tieneAceptado(envio.getUsuarioId(), envio.getRetoId());

            guardado.setVeredicto(veredicto);
            guardado = repo.save(guardado);

            if (esPrimerAceptado) {
                EnvioAceptadoEvent evento =
                        new EnvioAceptadoEvent(envio.getUsuarioId(), envio.getRetoId());
                for (EnvioObserver observer : observers) {
                    observer.onEnvioAceptado(evento);
                }
            }
        } catch (Exception e) {
            guardado.setVeredicto(Veredicto.ERROR_EN_EJECUCION);
            guardado = repo.save(guardado);
        }

        return guardado;
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<Envio> buscarPorId(long id) {
        return repo.findById(id);
    }

    @Override
    @Transactional(readOnly = true)
    public List<Envio> listarPorUsuario(long usuarioId) {
        return repo.findByUsuarioId(usuarioId);
    }

    @Override
    @Transactional(readOnly = true)
    public List<Envio> listarPorReto(long retoId) {
        return repo.findByRetoId(retoId);
    }

    @Override
    @Transactional(readOnly = true)
    public boolean tieneAceptado(long usuarioId, long retoId) {
        return repo.existsByUsuarioIdAndRetoIdAndVeredicto(usuarioId, retoId, Veredicto.ACEPTADO);
    }
}
